From 73804e4ba6a07f9effe9826ac6d3ca6585fa229b Mon Sep 17 00:00:00 2001
From: Gregory CLEMENT <gregory.clement@bootlin.com>
Date: Fri, 8 Oct 2021 14:49:48 +0200
Subject: [PATCH] brcmfmac: Allowing disabling 802.11ac feature

Actually it disables the VHT capabilities.

It allows to ensure that 802.11ac won't be used.

Signed-off-by: Gregory CLEMENT <gregory.clement@bootlin.com>
---
 .../net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c    | 7 ++++++-
 drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c  | 5 +++++
 drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h  | 1 +
 3 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
index 2a127440a7cf..8c1ed641a03d 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/cfg80211.c
@@ -7103,10 +7103,15 @@ static int brcmf_setup_wiphybands(struct brcmf_cfg80211_info *cfg)
 	} else {
 		brcmf_get_bwcap(ifp, bw_cap);
 	}
-	brcmf_dbg(INFO, "nmode=%d, vhtmode=%d, bw_cap=(%d, %d)\n",
+	brcmf_dbg(INFO, "nmode=%d, vhtmode=%d , bw_cap=(%d, %d)\n",
 		  nmode, vhtmode, bw_cap[NL80211_BAND_2GHZ],
 		  bw_cap[NL80211_BAND_5GHZ]);
 
+	if (drvr->settings->disable_11ac) {
+		vhtmode = 0;
+		brcmf_dbg(INFO, "vhtmode disabled by module parameter\n");
+	}
+
 	err = brcmf_fil_iovar_int_get(ifp, "rxchain", &rxchain);
 	if (err) {
 		bphy_err(drvr, "rxchain error (%d)\n", err);
diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
index 2825d6575dc1..9809a4175e25 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.c
@@ -82,6 +82,10 @@ static int brcmf_max_pm;
 module_param_named(max_pm, brcmf_max_pm, int, 0);
 MODULE_PARM_DESC(max_pm, "Use max power management mode by default");
 
+static int brcmf_disable_11ac;
+module_param_named(disable_11ac, brcmf_disable_11ac, int, 0);
+MODULE_PARM_DESC(disable_11ac, "Disable VHT capabilities(802.11ac)");
+
 #ifdef DEBUG
 /* always succeed brcmf_bus_started() */
 static int brcmf_ignore_probe_fail;
@@ -487,6 +491,7 @@ struct brcmf_mp_device *brcmf_get_module_param(struct device *dev,
 	settings->eap_restrict = !!brcmf_eap_restrict;
 	settings->sdio_wq_highpri = !!brcmf_sdio_wq_highpri;
 	settings->default_pm = !!brcmf_max_pm ? PM_MAX : PM_FAST;
+	settings->disable_11ac = !!brcmf_disable_11ac;
 #ifdef DEBUG
 	settings->ignore_probe_fail = !!brcmf_ignore_probe_fail;
 #endif
diff --git a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h
index c0bf5867af5f..fa8430bbe4dd 100644
--- a/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h
+++ b/drivers/net/wireless/broadcom/brcm80211/brcmfmac/common.h
@@ -54,6 +54,7 @@ struct brcmf_mp_device {
 	bool		sdio_wq_highpri;
 	int		default_pm;
 	bool		ignore_probe_fail;
+	bool		disable_11ac;
 	struct brcmfmac_pd_cc *country_codes;
 	const char	*board_type;
 	union {
-- 
2.33.0

